# DS-SF-36 | 04 | Databases and Scrapping | Assignment | Starter Code

## `SQLite` and Bistro

In this assignment, we will be exploring the `bistro` dataset.  The previous assignment used `pandas`.  Today, we'll answer the same questions but using `SQLite`.  In some situations, `pandas` will be a better solution.  In others, doing it using `SQL` will make more sense.  As you gain more experience, you'll know which one to use.

> ### Question 1.  Import the `sqlite3` package.

In [5]:
import os

import pandas as pd
pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

import sqlite3

> ### Question 2.  Connect to the `dataset-04-bistro.db` database.  The rest of this assignment focus on the `bistro` table.

In [6]:
db = sqlite3.connect(os.path.join('..', 'datasets', 'dataset-04-bistro.db'))

> ### Question 3.  How many samples (i.e., rows) are in this dataset?

In [7]:
pd.io.sql.read_sql(
'''
SELECT *
    FROM bistro

;
''', con = db)

index       day    time      name  gender  is_smoker  party  check   tip
0        0    Sunday  Dinner  Kimberly  Female          0      2  16.99  1.01
1        1    Sunday  Dinner  Nicholas    Male          0      3  10.34  1.66
2        2    Sunday  Dinner     Larry    Male          0      3  21.01  3.50
3        3    Sunday  Dinner    Joseph    Male          0      2  23.68  3.31
4        4    Sunday  Dinner    Janice  Female          0      4  24.59  3.61
..     ...       ...     ...       ...     ...        ...    ...    ...   ...
239    239  Saturday  Dinner     Kevin    Male          0      3  29.03  5.92
240    240  Saturday  Dinner    Sandra  Female          0      2  27.18  2.00
241    241  Saturday  Dinner      Carl    Male          0      2  22.67  2.00
242    242  Saturday  Dinner       Jon    Male          0      2  17.82  1.75
243    243  Thursday  Dinner    Brandi  Female          0      2  18.78  3.00

[244 rows x 9 columns]

Answer: 244 samples/rows.

In [11]:
pd.io.sql.read_sql(
'''
SELECT COUNT()
    FROM bistro

;
''', con = db)

COUNT()
0      244

> ### Question 4.  Print the first two rows of the table to the console.

In [8]:
pd.io.sql.read_sql(
'''
SELECT *
    FROM bistro
    LIMIT 2

;
''', con = db)

index     day    time      name  gender  is_smoker  party  check   tip
0      0  Sunday  Dinner  Kimberly  Female          0      2  16.99  1.01
1      1  Sunday  Dinner  Nicholas    Male          0      3  10.34  1.66

> ### Question 5.  For which week days does the dataset has data for?

In [12]:
pd.io.sql.read_sql(
'''
SELECT DISTINCT day
    FROM bistro
    ORDER BY day
;
''', con = db)

day
0    Friday
1  Saturday
2    Sunday
3  Thursday

> ### Question 6.  How often was the bistro patronized for each week day?

In [13]:
pd.io.sql.read_sql(
'''
SELECT day, COUNT() AS count
    FROM bistro
    GROUP BY day
;
''', con = db)

day  count
0    Friday     19
1  Saturday     87
2    Sunday     76
3  Thursday     62

Answer: Thursday, Friday, Saturday, Sunday.

> ### Question 7.  How much tip did waiters collect for each week day?

In [20]:
pd.io.sql.read_sql(
'''
SELECT day, SUM(tip) as total_tips

    FROM bistro
    GROUP BY day
;
''', con = db)

day  total_tips
0    Friday       51.96
1  Saturday      260.40
2    Sunday      247.39
3  Thursday      171.83

Answer: See table above for totals, if you are seeking the average per day, see below

In [23]:
pd.io.sql.read_sql(
'''
SELECT day, AVG(tip) as avg_tips

    FROM bistro
    GROUP BY day
    ORDER by day
;
''', con = db)

day  avg_tips
0    Friday  2.734737
1  Saturday  2.993103
2    Sunday  3.255132
3  Thursday  2.771452

> ### Question 8.  What is the average tip per check (in absolute \$) for each week day?

In [ ]:
# See above for my way to do it, or the answer key way below

In [22]:
pd.io.sql.read_sql(
'''
SELECT day, SUM(tip)/ COUNT() as avg_tip_per_check

    FROM bistro
    GROUP BY day
    ORDER BY day

;
''', con = db)

day  avg_tip_per_check
0    Friday           2.734737
1  Saturday           2.993103
2    Sunday           3.255132
3  Thursday           2.771452

Answer: TODO

> ### Question 9.  What is the average tip per check (as a percentage of the check) for each week day?

(`CHECK` is a reserved keywork; use `` `check` `` (put the name between backticks) to reference the `check` column)

In [26]:
pd.io.sql.read_sql(
'''
SELECT day, SUM(tip) / SUM(`check`) * 100 as avg_tip_per_check

    FROM bistro
    GROUP BY day
    ORDER BY day

;
''', con = db)

day  avg_tip_per_check
0    Friday          15.944519
1  Saturday          14.642375
2    Sunday          15.203791
3  Thursday          15.673201

Answer: See above.

> ### Question 10.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

In [28]:
pd.io.sql.read_sql(
'''
SELECT DISTINCT name
    FROM bistro
    WHERE gender == 'Male'
INTERSECT
SELECT DISTINCT name
    FROM bistro
    WHERE gender == 'Female'

;
''', con = db)

name
0  Casey

Answer: Yes, 1 name, Casey.

> ### Question 11.  If no patrons share the same name, how many unique patrons are in the dataset?

In [38]:
pd.io.sql.read_sql(
'''
WITH names_by_gender AS
    (SELECT name, gender
        FROM bistro
        GROUP BY name, gender)
        
SELECT COUNT()
    FROM names_by_gender
;
''', con = db)

COUNT()
0      182

In [ ]:
# proper way above, my original way below

In [31]:
pd.io.sql.read_sql(
'''
SELECT DISTINCT name
    FROM bistro

;
''', con = db)

name
0    Kimberly
1    Nicholas
2       Larry
3      Joseph
4      Janice
..        ...
176    Darwin
177     Henry
178    Jeremy
179   Dorothy
180    Sandra

[181 rows x 1 columns]

Answer: 181 patrons, presuming no one shares the same name.

> ### Question 12.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [46]:
pd.io.sql.read_sql(
'''
SELECT COUNT() AS count
    FROM bistro
    WHERE name == 'Kevin'

;
''', con = db)

count
0      4

Answer: Kevin patronized the bistro 4 times.

In [47]:
pd.io.sql.read_sql(
'''
SELECT COUNT() AS count
    FROM bistro
    WHERE name == 'Alice'

;
''', con = db)

count
0      2

> ### Question 13.  Who are the top 3 female and male patrons?

In [53]:
# top part for practice to make sure I understand

In [56]:
pd.io.sql.read_sql(
'''

WITH visits AS
    (SELECT name, gender, COUNT() as visits
        FROM bistro
        GROUP BY name, gender
        ORDER BY visits DESC)

SELECT * 
    FROM visits
    LIMIT 5
;
''', con = db)

name gender  visits
0  David   Male       8
1  Casey   Male       5
2  James   Male       5
3   Carl   Male       4
4  Kevin   Male       4

In [52]:
pd.io.sql.read_sql(
'''

WITH visits AS
    (SELECT name, gender, COUNT() as visits
        FROM bistro
        GROUP BY name, gender
        ORDER BY visits DESC),
    
top_female_patrons AS
    (Select *
        FROM visits
        WHERE gender == 'Female'
        LIMIT 3),
    
top_male_patrons AS
    (Select *
        FROM visits
        WHERE gender == 'Male'
        LIMIT 3)
    
SELECT * from top_female_patrons
UNION
SELECT * from top_male_patrons   
    
    
;
''', con = db)



name  gender  visits
0  Casey  Female       3
1  Casey    Male       5
2  David    Male       8
3  James    Male       5
4  Laura  Female       3
5   Mary  Female       4

Answer: TODO

> ### Question 14.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

In [63]:
pd.io.sql.read_sql(
'''

WITH best_tipper AS
    (SELECT name, SUM(tip) / SUM(`check`) * 100 as tips_per_check, COUNT() as times_patronizing_bistro
        FROM bistro
        GROUP BY name
        ORDER BY tips_per_check DESC)

SELECT *
    FROM best_tipper
    LIMIT 1
;
''', con = db)

name  tips_per_check  times_patronizing_bistro
0  Maryann       41.666667                         1

In [65]:
pd.io.sql.read_sql(
'''

WITH best_tipper AS
    (SELECT name, SUM(tip) / SUM(`check`) * 100 as tips_per_check, COUNT() as times_patronizing_bistro
        FROM bistro
        GROUP BY name
        ORDER BY tips_per_check)

SELECT *
    FROM best_tipper
    LIMIT 1
;
''', con = db)

name  tips_per_check  times_patronizing_bistro
0  Jeremy        3.563814                         1

Answer: Jeremy is the worst tipper.  He patronized once.

In [66]:
#redid their way below (I got the right answers above too)

In [68]:
pd.io.sql.read_sql(
'''

SELECT name, SUM(tip) / SUM(`check`) * 100 AS tips_per_check, COUNT() AS times_patronizing_bistro
    FROM bistro
    GROUP BY name
    ORDER BY tips_per_check DESC
    LIMIT 1

;
''', con = db)

name  tips_per_check  times_patronizing_bistro
0  Maryann       41.666667                         1

In [69]:
pd.io.sql.read_sql(
'''

SELECT name, SUM(tip) / SUM(`check`) * 100 AS tips_per_check, COUNT() AS times_patronizing_bistro
    FROM bistro
    GROUP BY name
    ORDER BY tips_per_check
    LIMIT 1

;
''', con = db)

name  tips_per_check  times_patronizing_bistro
0  Jeremy        3.563814                         1